rfcs/20211118-support-implicit-concat/README.md

```diff
@@ -0,0 +1,132 @@
+# RFC: Support of implicit concat
+
+## Introduction
+
+This document explores a way of supporting implicit concat. Implicit concat is a
+concat primitive which is done without any additional copies just by storing
+correspondent parts in the right locations of the final tensor. Such approach
+saves at least one memory pass over the final output. Sometimes it may be two or
+more.
+
+### Motivation
+
+Because compute power tends to increase more quickly than the memory bandwidth
+needed to support it, most optimizations target saving work with memory and move
+more in a compute field. Same desire here. We want to use intelligence to save
+on putting smaller memories into a bigger one just with right format.
+
+## Proposal
+
+Let's start with the fact that such optimization cannot be provided by oneDNN
+on API level - it must be enabled on user side manually due to several reasons,
+primarily because of memory management.
+
+Let's consider them on an example which includes only two tensors to create a
+bigger one:
+* It may happen that two inputs, being an output from other primitives, may have
+  different memory layout which are incompatible, e.g., aBx32b and ABx16a32b. In
+  this case reorder would be needed no matter what, thus, doesn't make sense to
+  avoid explicit concat which would execute such reorder as a part of operation.
+  It means that user oversees that formats are same or compatible for implicit
+  merging. Note that this holds for each architecture and even if input memory
+  formats are same for one architecture, it does not necessarily mean that the
+  other one would have exact format or match would happen.
+* But even though formats are same, it may happen that implicit concat is still
+  impossible, e.g., first input has padding. Despite in theory implicit concat
+  is still possible, in the reality it may bring only troubles since second
+  input would highly likely be unaligned in memory and may lead to various
+  "interesting" performance effects or, even, crashes, so that using explicit
+  concat would be at least safer, but at most more beneficial.
```

Comment:
Since safety is kind of a paramount virtue, it's confusing to say "at least safer". Also, "more beneficial" is vague--more beneficial in what way? I suggest "would be safer, and potentially offer better performance."

Reply:
Thanks for a suggestion, applied.